std/statet.glu

```diff
@@ -0,0 +1,83 @@
+//! The state monad transformer.
+
+let { Applicative, Functor, Monad } = import! std.prelude
+let { Alternative, (<|>), empty } = import! std.alternative
+let { wrap, (<*>) } = import! std.applicative
+let { (>>), (<<), ? } = import! std.function
+let { map } = import! std.functor
+let { (>>=) } = import! std.monad
+let { Transformer } = import! std.transformer
+
+type StateOut s a = { value : a, state : s }
+type WrStateOut s m a = m { value : a, state : s }
+
+type StateT s m a = s -> m { value : a, state : s }
+
+let map_sout f st : (a -> b) -> StateOut s a -> StateOut s b =
+    {value = f st.value, state = st.state}
+
+let functor ?fr : [Functor m] -> Functor (StateT s m) =
+    let map f sr : (a -> b) -> StateT s m a -> StateT s m b =
+        fr.map (map_sout f) << sr
+
+    { map }
+
+let applicative ?mo : [Monad m] -> Applicative (StateT s m) =
+    let apply srf sr : StateT s m (a -> b) -> StateT s m a -> StateT s m b = \state ->
+        srf state >>= \fout ->
+            let {value = f, state = state'} = fout
+            mo.applicative.functor.map (map_sout f) (sr state')
+
+    let wrap value : a -> StateT s m a = \state ->
+        mo.applicative.wrap { value, state }
+
+    { functor = functor ?mo.applicative.functor, apply, wrap }
+
+let monad ?mo : [Monad m] -> Monad (StateT s m) =
+    let flat_map f sr : (a -> StateT s m b) -> StateT s m a -> StateT s m b = \state ->
+        sr state >>= \sout ->
+            let {value, state = state'} = sout
+            f value state'
+
+    { applicative, flat_map }
+
+// FIXME implicit parameter troubles
```

Comment:
I assume that this is the error that occur on appveyor https://ci.appveyor.com/project/Marwes/gluon/builds/22319979#L870 ? Will look into it.

Reply:
Yep. Those are the errors I'm getting on my machine. Thanks for taking a look; I wasn't able to make sense of them.

Reply:
#689 takes care of the second error